driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+        }
+
+        ServerSelector apply(final ServerSelector selector) {
+            return new DeprioritizingSelector(selector);
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
+        }
+
+        public void onAttemptFailure(final Throwable failure) {
+            if (!(failure instanceof MongoConnectionPoolClearedException)) {
+                deprioritized.add(candidate);
```

Comment:
Thanks! There was actually a bug (though, not observable): `candidate` is allowed to be `null` at this point, but `onAttemptFailure` must not add it to `deprioritized`. Fortunately for us, `HashSet` allows `null` values, so it was fine. Nevertheless, I [fixed the method and added a test](https://github.com/mongodb/mongo-java-driver/pull/1367/commits/7cba88b2fab9027d7eab35ed8dfdaddaf4d5a868), though the test passes both before and after the change, again, because of `HashSet`.